### PR TITLE
Seek issue

### PR DESCRIPTION
**Pull request by Yves33**
This PR introduces proper seeking in VALI
+ in current state, seeking decodes all frames from start of video, resulting in long latency when seeking at large offsets (specially if resolution is high).
+ with PR, VALI seeks to previous keyframe and decodes all frames till it reaches the correct timestamp. As a result, the number of decoded frames should never be higher than GOP

the PR has been tested with fake (ffmpeg generated) and real mp4 h264 videos.

### CODE REVIEW

File: src/TC/src/TaskDecodeFrame.cpp [hunk 1]
```diff
@@ -605,7 +605,7 @@ struct FfmpegDecodeFrame_Impl {
          * To deal with that, discard PTS and make libavcodec come up with
          * correct auto-generated value.
          */
-        pkt->pts = AV_NOPTS_VALUE;
+        //pkt->pts = AV_NOPTS_VALUE; // What is that line used for? It prevents proper seeking
```
Comment (RomanArzumanyan) [inline src/TC/src/TaskDecodeFrame.cpp:608]: There's a clear comment above this line that explains it's presence.

**Comment by RomanArzumanyan** [comment]
Hi @Yves33 

Does this MR pass the unit tests locally on your machine ?

**Comment by Yves33** [comment]
Yes, I saw the clear comment above this line , but does not seem necessary in my tests, and prevents seeking (as stated in the clear comment on the side of this line :-)
(If that line is required, then we should find a way to restore correct pts)

regarding unittests:  
+ All unittests pass (at the price of changing  pynvml to py3nvml in test_common.py - old machine here!)
+ Could not test torch dependant ones PySurface and TorchSegmentation - don't have torch on my machine!)

**Comment by RomanArzumanyan** [comment]
@Yves33 

Please take a look at this comment:
https://github.com/RomanArzumanyan/VALI/issues/74#issuecomment-2247643163

IIRC pts restore was required to fix non-monotonous pts increase.
Can you test your PR on that video ?

**Comment by Yves33** [comment]
well, there are many videos in that folder...
But using "pvd.mp4" as test video, the frame decoded by VALI at 127.60s really looks identical to the frame decoded by avidemux at 2'07"630.  Looks the same for other timestamps (visually) checked.

**Comment by Yves33** [comment]
one additional comment:
SeekContext has a num_frames_decoded and out_frame_pts fields. these are not updated by current PR.
Should we and How to return them to python (they are marked readonly)?

**Comment by Yves33** [comment]
Is there a way to (and should we?)  update ctx.num_frames_decoded and ctx.out_frame_pts in SeekDecode() and pass these values back to Python?

**Comment by Yves33** [comment]
Strange behavior: last commit (break decoding if SeekContext.mode==PREV_KEY_FRAME breaks PyDecoder unittest !
from my understanding, this should be the revese : 
+ previously: breaking on EXACT_FRAME, so we always stopped at keyframe (but unittest used to pass, strangely!)
+ latest commit: breaking on PREV_KEY_FRAME, so we decode till appropriate timestamp (but unittest fails!)

File: src/TC/src/TaskDecodeFrame.cpp (hunk 1 above)
Comment (Yves33) [reply src/TC/src/TaskDecodeFrame.cpp:608]: Yes, I saw the clear comment above this line , but does not seem necessary in my tests, and prevents seeking (as stated in the clear comment on the side of this line :-)
(If that line is required, then we should find a way to restore correct pts)

**Comment by sandhawalia** [comment]
hey @RomanArzumanyan glad to see you are continuing the great work from VPF. I had a bunch of PRs in the early days VPF (still being used btw) and now we are planning to include HW accelerated video decoding with VALI in [rbyte](https://github.com/yaak-ai/rbyte). This issue is currently blocking us. Would be amazing to have this merged.

**Comment by RomanArzumanyan** [comment]
Hi @sandhawalia 

Nice to meet you again!

I’m currently on vacation, will come back to this topic with 2 weeks. Before leaving for the vacation, I have finally found the video that caused non monotonous pts increase and will add it to test database.

Then I’ll test this PR in main repo feature branch. Unfortunately my self-hosted runner environment doesn’t work in PRs from other people’s fork for some reason.

Im also looking forward to get this PR merged!

**Pull request merged**